Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift

```diff
@@ -0,0 +1,88 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Request to test the validity of a rename operation at the given location.
+///
+/// Looks up the symbol at the given position and returns the range of the symbol the user is renaming,
+/// or null if there is no symbol to be renamed at the given position
+///
+/// Servers that provide rename preparation should set`prepareProvider` to true in the `renameProvider` server capability.
+///
+/// - Parameters:
+///   - textDocument: The document in which to lookup the symbol location.
+///   - position: The document location at which to lookup symbol information.
+///
+/// - Returns: A range for the symbol, and optionally a placeholder text of the string content to be renamed.
+public struct PrepareRenameRequest: TextDocumentRequest, Hashable {
+  public static let method: String = "textDocument/prepareRename"
+  public typealias Response = PrepareRenameResponse?
+
+  /// The document in which to lookup the symbol location.
+  public var textDocument: TextDocumentIdentifier
+
+  /// The document location at which to lookup symbol information.
+  public var position: Position
+
+  public init(textDocument: TextDocumentIdentifier, position: Position) {
+    self.textDocument = textDocument
+    self.position = position
+  }
+}
+
+public struct PrepareRenameResponse: ResponseType, Hashable {
```

Comment:
Looks like in the spec it's defined as `Range | { range: Range, placeholder: string } | { defaultBehavior: boolean } | null` - do we need to handle the defaultBehavior case too?

Reply:
defaultBehavior was added in LSP 3.16, and as I said, I’m not supporting that (would require adding it to the client capability, adding the appropriate rename option for what the default behavior is, etc). IMO it isn’t necessary

Reply:
I think it’s interesting that LSP added a new response option with a new version of LSP. AFAICT the server can’t know whether the client supports this response option or am I missing something?

---

In any case, wouldn’t it be cleaner to explicitly model the three different response types by e.g. making `PrepareRenameResponse` as an `enum`? As it stands now, we use `placeholder == nil` as a flag to convey whether the response was a raw `Range` or `{ range: Range, placeholder: string }`. This means that 
```json
{
  "range": …,
  "placeholder": null
}
```
wouldn’t round-trip. I know, the spec requires `placeholder` to be non-optional but still, we might as well prepare for `defaultBehavior` now.

Reply:
Client Capabilities have an additional `renameProvider.prepareSupportDefaultBehavior` property to specify support

Reply:
Ah, that makes sense. I’d be fine to merge this as-is.